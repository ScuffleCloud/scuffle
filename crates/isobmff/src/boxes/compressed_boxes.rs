use scuffle_bytes_util::BytesCow;

use super::{ExtendedTypeBox, MovieBox, MovieFragmentBox, SegmentIndexBox, SubsegmentIndexBox};
use crate::{IsoBox, UnknownBox};

/// Original file-type box
///
/// ISO/IEC 14496-12 - 8.19.5
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"otyp", crate_path = crate)]
pub struct OriginalFileTypeBox<'a> {
    #[iso_box(nested_box(collect))]
    pub etyp: Option<ExtendedTypeBox<'a>>,
    #[iso_box(nested_box(collect_unknown))]
    pub unknown_boxes: Vec<UnknownBox<'a>>,
}

pub trait CompressedBox {
    type UncompressedBox: IsoBox;
}

/// Compressed movie box
///
/// ISO/IEC 14496-12 - 8.19.6
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"!mov", crate_path = crate)]
pub struct CompressedMovieBox<'a> {
    pub data: BytesCow<'a>,
}

impl<'a> CompressedBox for CompressedMovieBox<'a> {
    type UncompressedBox = MovieBox<'a>;
}

/// Compressed movie fragment box
///
/// ISO/IEC 14496-12 - 8.19.7
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"!mof", crate_path = crate)]
pub struct CompressedMovieFragmentBox<'a> {
    pub data: BytesCow<'a>,
}

impl<'a> CompressedBox for CompressedMovieFragmentBox<'a> {
    type UncompressedBox = MovieFragmentBox<'a>;
}

/// Compressed segment index box
///
/// ISO/IEC 14496-12 - 8.19.8
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"!six", crate_path = crate)]
pub struct CompressedSegmentIndexBox<'a> {
    pub data: BytesCow<'a>,
}

impl CompressedBox for CompressedSegmentIndexBox<'_> {
    type UncompressedBox = SegmentIndexBox;
}

/// Compressed subsegment index box
///
/// ISO/IEC 14496-12 - 8.19.9
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"!ssx", crate_path = crate)]
pub struct CompressedSubsegmentIndexBox<'a> {
    pub data: BytesCow<'a>,
}

impl CompressedBox for CompressedSubsegmentIndexBox<'_> {
    type UncompressedBox = SubsegmentIndexBox;
}
