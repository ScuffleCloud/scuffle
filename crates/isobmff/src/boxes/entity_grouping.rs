use scuffle_bytes_util::zero_copy::{Deserialize, Serialize};

use crate::{BoxHeader, FullBoxHeader, IsoBox, IsoSized, UnknownBox};

/// Groups list box
///
/// ISO/IEC 14496-12 - 8.18.2
#[derive(IsoBox, Debug)]
#[iso_box(box_type = b"grpl", crate_path = crate)]
pub struct GroupsListBox<'a> {
    #[iso_box(header)]
    pub header: BoxHeader,
    #[iso_box(nested_box(collect))]
    pub altr: Vec<AltrEntityToGroupBox>,
    #[iso_box(nested_box(collect_unknown))]
    pub unknown_boxes: Vec<UnknownBox<'a>>,
}

/// Entity to group box
///
/// ISO/IEC 14496-12 - 8.18.3
#[derive(Debug)]
pub struct EntityToGroupBox {
    pub group_id: u32,
    pub num_entities_in_group: u32,
    pub entity_id: Vec<u32>,
}

impl<'a> Deserialize<'a> for EntityToGroupBox {
    fn deserialize<R>(mut reader: R) -> std::io::Result<Self>
    where
        R: scuffle_bytes_util::zero_copy::ZeroCopyReader<'a>,
    {
        let group_id = u32::deserialize(&mut reader)?;
        let num_entities_in_group = u32::deserialize(&mut reader)?;

        let mut entity_id = Vec::with_capacity(num_entities_in_group as usize);
        for _ in 0..num_entities_in_group {
            entity_id.push(u32::deserialize(&mut reader)?);
        }

        Ok(EntityToGroupBox {
            group_id,
            num_entities_in_group,
            entity_id,
        })
    }
}

impl Serialize for EntityToGroupBox {
    fn serialize<W>(&self, mut writer: W) -> std::io::Result<()>
    where
        W: std::io::Write,
    {
        self.group_id.serialize(&mut writer)?;
        self.num_entities_in_group.serialize(&mut writer)?;

        for id in &self.entity_id {
            id.serialize(&mut writer)?;
        }

        Ok(())
    }
}

impl IsoSized for EntityToGroupBox {
    fn size(&self) -> usize {
        self.group_id.size() + self.num_entities_in_group.size() + self.entity_id.size()
    }
}

#[derive(IsoBox, Debug)]
#[iso_box(box_type = b"altr", crate_path = crate)]
pub struct AltrEntityToGroupBox {
    #[iso_box(header)]
    pub header: FullBoxHeader,
    pub entity_to_group: EntityToGroupBox,
}
