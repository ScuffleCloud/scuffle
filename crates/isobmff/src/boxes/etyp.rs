use super::Tyco;
use crate::{BoxHeader, IsoBox};

/// Extended type box
///
/// ISO/IEC 14496-12 - 4.4
#[derive(IsoBox)]
#[iso_box(box_type = b"etyp", crate_path = "crate")]
pub struct Etyp {
    #[iso_box(header)]
    pub header: BoxHeader,
    #[iso_box(remaining)]
    pub compatible_combinations: Vec<Tyco>,
}
