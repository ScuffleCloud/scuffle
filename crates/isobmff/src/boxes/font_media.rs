use scuffle_bytes_util::zero_copy::{Deserialize, Serialize};

use super::SampleEntry;
use crate::IsoSized;

/// Font sample entry
///
/// ISO/IEC 14496-12 - 12.7.3
///
/// Sub boxes:
/// - [`btrt`](super::BitRateBox)
/// - Any other boxes
#[derive(Debug)]
pub struct FontSampleEntry {
    pub sample_entry: SampleEntry,
}

impl<'a> Deserialize<'a> for FontSampleEntry {
    fn deserialize<R>(reader: R) -> std::io::Result<Self>
    where
        R: scuffle_bytes_util::zero_copy::ZeroCopyReader<'a>,
    {
        Ok(Self {
            sample_entry: SampleEntry::deserialize(reader)?,
        })
    }
}

impl Serialize for FontSampleEntry {
    fn serialize<W>(&self, writer: W) -> std::io::Result<()>
    where
        W: std::io::Write,
    {
        self.sample_entry.serialize(writer)
    }
}

impl IsoSized for FontSampleEntry {
    fn size(&self) -> usize {
        self.sample_entry.size()
    }
}
