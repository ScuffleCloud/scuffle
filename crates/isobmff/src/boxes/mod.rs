mod ftyp;
pub use ftyp::*;

mod etyp;
pub use etyp::*;

mod tyco;
pub use tyco::*;

mod mdat;
pub use mdat::*;

mod free;
pub use free::*;
