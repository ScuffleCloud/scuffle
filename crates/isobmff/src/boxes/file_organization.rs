//! File organization boxes defined in ISO/IEC 14496-12 - 4

use nutype_enum::nutype_enum;

use crate::{IsoBox, IsoSized, UnknownBox};

nutype_enum! {
    /// A four character code, registered with ISO, that identifies a precise specification.
    ///
    /// See: https://mp4ra.org/registered-types/brands
    pub enum Brand([u8; 4]) {
        Mp41 = *b"mp41",
        IsoM = *b"isom",
        Avc1 = *b"avc1",
        Iso2 = *b"iso2",
        Mp71 = *b"mp71",
        Iso3 = *b"iso3",
        Iso4 = *b"iso4",
        Iso5 = *b"iso5",
        Iso6 = *b"iso6",
        Iso7 = *b"iso7",
        Iso8 = *b"iso8",
        Iso9 = *b"iso9",
        IsoA = *b"isoa",
        IsoB = *b"isob",
        Relo = *b"relo",
        IsoC = *b"isoc",
        Comp = *b"comp",
        Unif = *b"unif",
    }
}

impl IsoSized for Brand {
    fn size(&self) -> usize {
        4
    }
}

/// File-type box
///
/// ISO/IEC 14496-12 - 4.3
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"ftyp", crate_path = crate)]
pub struct FileTypeBox {
    /// The "best use" brand of the file which will provide the greatest compatibility.
    #[iso_box(from = "[u8; 4]")]
    pub major_brand: Brand,
    /// Minor version of the major brand.
    pub minor_version: u32,
    /// A list of compatible brands.
    #[iso_box(repeated, from = "[u8; 4]")]
    pub compatible_brands: Vec<Brand>,
}

/// Type combination box
///
/// ISO/IEC 14496-12 - 4.4
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"tyco", crate_path = crate)]
pub struct TypeCombinationBox {
    /// A list of compatible brands.
    #[iso_box(repeated, from = "[u8; 4]")]
    pub compatible_brands: Vec<Brand>,
}

/// Extended type box
///
/// ISO/IEC 14496-12 - 4.4
#[derive(IsoBox, Debug, PartialEq, Eq)]
#[iso_box(box_type = b"etyp", crate_path = crate)]
pub struct ExtendedTypeBox<'a> {
    /// A list of [`TypeCombinationBox`]es.
    #[iso_box(nested_box(collect))]
    pub compatible_combinations: Vec<TypeCombinationBox>,
    /// A list of unknown boxes that were encountered while parsing the box.
    #[iso_box(nested_box(collect_unknown))]
    pub unknown_boxes: Vec<UnknownBox<'a>>,
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use scuffle_bytes_util::zero_copy::{Deserialize, Slice};

    use crate::boxes::{Brand, ExtendedTypeBox, TypeCombinationBox};

    #[test]
    fn demux_tyco() {
        #[rustfmt::skip]
        let data = [
            0x00, 0x00, 0x00, 0x0C, // size
            b't', b'y', b'c', b'o', // type
            b'i', b's', b'o', b'6', // data
            0x01,
        ];

        let mdat = TypeCombinationBox::deserialize(Slice::from(&data[..])).unwrap();
        assert_eq!(mdat.compatible_brands.len(), 1);
        assert_eq!(mdat.compatible_brands[0], Brand::Iso6);
    }

    #[test]
    fn demux_etyp() {
        #[rustfmt::skip]
        let data = [
            0x00, 0x00, 0x00, 44, // size
            b'e', b't', b'y', b'p', // type

            0x00, 0x00, 0x00, 12, // tyco size
            b't', b'y', b'c', b'o', // tyco type
            b'i', b's', b'o', b'm', // data

            0x00, 0x00, 0x00, 12, // tyco size
            b't', b'y', b'c', b'o', // tyco type
            b'i', b's', b'o', b'6', // data

            0x00, 0x00, 0x00, 12, // unknown size
            b'u', b'n', b'k', b'n', // unknown type
            0x42, 0x00, 0x42, 0x00, // data
        ];

        let mdat = ExtendedTypeBox::deserialize(Slice::from(&data[..])).unwrap();

        assert_eq!(mdat.compatible_combinations.len(), 2);

        assert_eq!(mdat.compatible_combinations[0].compatible_brands.len(), 1);
        assert_eq!(mdat.compatible_combinations[0].compatible_brands[0], Brand::IsoM);

        assert_eq!(mdat.compatible_combinations[1].compatible_brands.len(), 1);
        assert_eq!(mdat.compatible_combinations[1].compatible_brands[0], Brand::Iso6);

        assert_eq!(mdat.unknown_boxes.len(), 1);
        assert_eq!(mdat.unknown_boxes[0].data.len(), 4);
        assert_eq!(mdat.unknown_boxes[0].data.as_bytes()[0], 0x42);
        assert_eq!(mdat.unknown_boxes[0].data.as_bytes()[1], 0x00);
        assert_eq!(mdat.unknown_boxes[0].data.as_bytes()[2], 0x42);
        assert_eq!(mdat.unknown_boxes[0].data.as_bytes()[3], 0x00);
    }
}
