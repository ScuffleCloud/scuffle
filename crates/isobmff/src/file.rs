use scuffle_bytes_util::zero_copy::{Deserialize, DeserializeSeed};

use crate::boxes::{
    ExtendedTypeBox, FileTypeBox, IdentifiedMediaDataBox, MediaDataBox, MetaBox, MovieBox, MovieFragmentBox, ProgressiveDownloadInfoBox
};
use crate::{IsoBox, UnknownBox};

#[derive(IsoBox, Debug)]
#[iso_box(box_type = b"root", skip_deserialize_impl, crate_path = crate)] // The box type does not matter here
pub struct IsobmffFile<'a> {
    #[iso_box(header)]
    _empty_header: (),
    pub ftyp: FileTypeBox,
    #[iso_box(nested_box(collect))]
    pub etyp: Vec<ExtendedTypeBox<'a>>,
    #[iso_box(nested_box(collect))]
    pub mdat: Vec<MediaDataBox<'a>>,
    #[iso_box(nested_box(collect))]
    pub pdin: Option<ProgressiveDownloadInfoBox>,
    #[iso_box(nested_box(collect))]
    pub imda: Vec<IdentifiedMediaDataBox<'a>>,
    #[iso_box(nested_box)]
    pub moov: MovieBox<'a>,
    #[iso_box(nested_box(collect))]
    pub moof: Vec<MovieFragmentBox<'a>>,
    // pub mfra: Vec<MovieFragmentRandomAccessBox>,
    #[iso_box(nested_box(collect))]
    pub meta: Option<MetaBox<'a>>,
    // pub styp: Vec<SegmentTypeBox>,
    // pub sidx: Vec<SegmentIndexBox>,
    // pub ssix: Vec<SubsegmentIndexBox>,
    // pub prft: Vec<ProducerReferenceTimeBox>,
    // pub otyp: Vec<OriginalFileTypeBox>,
    #[iso_box(nested_box(collect_unknown))]
    pub unknown_boxes: Vec<UnknownBox<'a>>,
}

impl<'a> Deserialize<'a> for IsobmffFile<'a> {
    fn deserialize<R>(reader: R) -> std::io::Result<Self>
    where
        R: scuffle_bytes_util::zero_copy::ZeroCopyReader<'a>,
    {
        <Self as DeserializeSeed<()>>::deserialize_seed(reader, ())
    }
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use std::path::PathBuf;

    use scuffle_bytes_util::zero_copy::Deserialize;

    use super::IsobmffFile;

    #[test]
    fn avc_aac_sample() {
        let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../assets");
        let data = std::fs::read(dir.join("avc_aac.mp4").to_str().unwrap()).unwrap();
        let mut reader = scuffle_bytes_util::zero_copy::Slice::from(&data[..]);
        let file = IsobmffFile::deserialize(&mut reader).unwrap();
        insta::assert_debug_snapshot!(file);
    }
}
