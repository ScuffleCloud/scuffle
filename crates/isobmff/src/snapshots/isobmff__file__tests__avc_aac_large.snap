---
source: crates/isobmff/src/file.rs
expression: og_file
snapshot_kind: text
---
IsobmffFile {
    ftyp: FileTypeBox {
        major_brand: Brand::IsoM,
        minor_version: 512,
        compatible_brands: [
            Brand::IsoM,
            Brand::Iso2,
            Brand::Avc1,
            Brand::Mp41,
        ],
    },
    etyp: [],
    otyp: [],
    pdin: None,
    moov: MovieBox {
        mvhd: MovieHeaderBox {
            full_header: FullBoxHeader {
                version: 0,
                flags: U24Be(
                    0,
                ),
            },
            creation_time: 0,
            modification_time: 0,
            timescale: 1000,
            duration: 1067,
            rate: 1,
            volume: 1,
            reserved1: 0,
            reserved2: 0,
            matrix: [
                65536,
                0,
                0,
                0,
                65536,
                0,
                0,
                0,
                1073741824,
            ],
            pre_defined: [
                0,
                0,
                0,
                0,
                0,
                0,
            ],
            next_track_id: 3,
        },
        meta: None,
        trak: [
            TrackBox {
                tkhd: TrackHeaderBox {
                    version: 0,
                    flags: TrackHeaderBoxFlags(
                        TrackEnabled | TrackInMovie,
                    ),
                    creation_time: 0,
                    modification_time: 0,
                    track_id: 1,
                    reserved1: 0,
                    duration: 1067,
                    reserved2: 0,
                    layer: 0,
                    alternate_group: 0,
                    volume: 0,
                    reserved3: 0,
                    matrix: [
                        65536,
                        0,
                        0,
                        0,
                        65536,
                        0,
                        0,
                        0,
                        1073741824,
                    ],
                    width: 3840,
                    height: 2160,
                },
                tref: None,
                trgr: None,
                edts: Some(
                    EditBox {
                        elst: Some(
                            EditListBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                entries: [
                                    EditListBoxEntry {
                                        edit_duration: 1067,
                                        media_time: 512,
                                        media_rate: 1,
                                    },
                                ],
                            },
                        ),
                    },
                ),
                ttyp: None,
                meta: None,
                mdia: MediaBox {
                    mdhd: MediaHeaderBox {
                        full_header: FullBoxHeader {
                            version: 0,
                            flags: U24Be(
                                0,
                            ),
                        },
                        creation_time: 0,
                        modification_time: 0,
                        timescale: 15360,
                        duration: 16384,
                        language: Langauge(
                            [
                                21,
                                14,
                                4,
                            ],
                        ),
                        pre_defined: 0,
                    },
                    hdlr: HandlerBox {
                        full_header: FullBoxHeader {
                            version: 0,
                            flags: U24Be(
                                0,
                            ),
                        },
                        pre_defined: 0,
                        handler_type: HandlerType::Video,
                        reserved: [
                            0,
                            0,
                            0,
                        ],
                        name: Utf8String(
                            "GPAC ISO Video Handler",
                        ),
                    },
                    elng: None,
                    minf: MediaInformationBox {
                        vmhd: Some(
                            VideoMediaHeaderBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        1,
                                    ),
                                },
                                graphicsmode: 0,
                                opcolor: [
                                    0,
                                    0,
                                    0,
                                ],
                            },
                        ),
                        smhd: None,
                        hmhd: None,
                        sthd: None,
                        vvhd: None,
                        unknown_boxes: [],
                        dinf: DataInformationBox {
                            dref: DataReferenceBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                url: [
                                    DataEntryUrlBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                1,
                                            ),
                                        },
                                        location: None,
                                    },
                                ],
                                urn: [],
                                imda: [],
                                snim: [],
                                unknown_boxes: [],
                            },
                        },
                        stbl: SampleTableBox {
                            stsd: SampleDescriptionBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                boxes: [
                                    UnknownBox {
                                        header: BoxHeader {
                                            size: Short(
                                                177,
                                            ),
                                            box_type: FourCc(
                                                "avc1",
                                            ),
                                        },
                                        data.len: 169,
                                    },
                                ],
                            },
                            stts: TimeToSampleBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                entries: [
                                    TimeToSampleBoxEntry {
                                        sample_count: 64,
                                        sample_delta: 256,
                                    },
                                ],
                            },
                            ctts: Some(
                                CompositionOffsetBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 60,
                                    entries.len: 60,
                                },
                            ),
                            cslg: None,
                            stsc: SampleToChunkBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 28,
                                entries.len: 28,
                            },
                            stsz: Some(
                                SampleSizeBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    sample_size: 0,
                                    sample_count: 64,
                                    entry_size.len: 64,
                                },
                            ),
                            stz2: None,
                            stco: Some(
                                ChunkOffsetBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 49,
                                    chunk_offset.len: 49,
                                },
                            ),
                            co64: None,
                            stss: Some(
                                SyncSampleBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 3,
                                    sample_number: [
                                        1,
                                        31,
                                        61,
                                    ],
                                },
                            ),
                            stsh: None,
                            padb: None,
                            stdp: None,
                            sdtp: None,
                            sbgp: [],
                            sgpd: [],
                            subs: [],
                            saiz: [],
                            saio: [],
                            csgp: [],
                        },
                    },
                    unknown_boxes: [],
                },
                unknown_boxes: [],
                udta: None,
            },
            TrackBox {
                tkhd: TrackHeaderBox {
                    version: 0,
                    flags: TrackHeaderBoxFlags(
                        TrackEnabled | TrackInMovie,
                    ),
                    creation_time: 0,
                    modification_time: 0,
                    track_id: 2,
                    reserved1: 0,
                    duration: 1003,
                    reserved2: 0,
                    layer: 0,
                    alternate_group: 1,
                    volume: 1,
                    reserved3: 0,
                    matrix: [
                        65536,
                        0,
                        0,
                        0,
                        65536,
                        0,
                        0,
                        0,
                        1073741824,
                    ],
                    width: 0,
                    height: 0,
                },
                tref: None,
                trgr: None,
                edts: Some(
                    EditBox {
                        elst: Some(
                            EditListBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                entries: [
                                    EditListBoxEntry {
                                        edit_duration: 1002,
                                        media_time: 1024,
                                        media_rate: 1,
                                    },
                                ],
                            },
                        ),
                    },
                ),
                ttyp: None,
                meta: None,
                mdia: MediaBox {
                    mdhd: MediaHeaderBox {
                        full_header: FullBoxHeader {
                            version: 0,
                            flags: U24Be(
                                0,
                            ),
                        },
                        creation_time: 0,
                        modification_time: 0,
                        timescale: 48000,
                        duration: 49152,
                        language: Langauge(
                            [
                                21,
                                14,
                                4,
                            ],
                        ),
                        pre_defined: 0,
                    },
                    hdlr: HandlerBox {
                        full_header: FullBoxHeader {
                            version: 0,
                            flags: U24Be(
                                0,
                            ),
                        },
                        pre_defined: 0,
                        handler_type: HandlerType::Audio,
                        reserved: [
                            0,
                            0,
                            0,
                        ],
                        name: Utf8String(
                            "GPAC ISO Audio Handler",
                        ),
                    },
                    elng: None,
                    minf: MediaInformationBox {
                        vmhd: None,
                        smhd: Some(
                            SoundMediaHeaderBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                balance: 0,
                                reserved: 0,
                            },
                        ),
                        hmhd: None,
                        sthd: None,
                        vvhd: None,
                        unknown_boxes: [],
                        dinf: DataInformationBox {
                            dref: DataReferenceBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                url: [
                                    DataEntryUrlBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                1,
                                            ),
                                        },
                                        location: None,
                                    },
                                ],
                                urn: [],
                                imda: [],
                                snim: [],
                                unknown_boxes: [],
                            },
                        },
                        stbl: SampleTableBox {
                            stsd: SampleDescriptionBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                boxes: [
                                    UnknownBox {
                                        header: BoxHeader {
                                            size: Short(
                                                110,
                                            ),
                                            box_type: FourCc(
                                                "mp4a",
                                            ),
                                        },
                                        data.len: 102,
                                    },
                                ],
                            },
                            stts: TimeToSampleBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                entries: [
                                    TimeToSampleBoxEntry {
                                        sample_count: 48,
                                        sample_delta: 1024,
                                    },
                                ],
                            },
                            ctts: None,
                            cslg: None,
                            stsc: SampleToChunkBox {
                                full_header: FullBoxHeader {
                                    version: 0,
                                    flags: U24Be(
                                        0,
                                    ),
                                },
                                entry_count: 1,
                                entries.len: 1,
                            },
                            stsz: Some(
                                SampleSizeBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    sample_size: 0,
                                    sample_count: 48,
                                    entry_size.len: 48,
                                },
                            ),
                            stz2: None,
                            stco: Some(
                                ChunkOffsetBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 48,
                                    chunk_offset.len: 48,
                                },
                            ),
                            co64: None,
                            stss: None,
                            stsh: None,
                            padb: None,
                            stdp: None,
                            sdtp: None,
                            sbgp: [
                                SampleToGroupBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    grouping_type: [
                                        114,
                                        111,
                                        108,
                                        108,
                                    ],
                                    grouping_type_parameter: None,
                                    entry_count: 1,
                                    entries: [
                                        SampleToGroupBoxEntry {
                                            sample_count: 48,
                                            group_description_index: 1,
                                        },
                                    ],
                                },
                            ],
                            sgpd: [
                                SampleGroupDescriptionBox {
                                    full_header: FullBoxHeader {
                                        version: 1,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    grouping_type: [
                                        114,
                                        111,
                                        108,
                                        108,
                                    ],
                                    default_length: Some(
                                        2,
                                    ),
                                    default_group_description_index: None,
                                    entry_count: 1,
                                    entries: [
                                        SampleGroupDescriptionEntry {
                                            description_length: None,
                                            sample_group_description_entry: RollRecoveryEntry {
                                                roll_distance: -1,
                                            },
                                        },
                                    ],
                                },
                            ],
                            subs: [],
                            saiz: [],
                            saio: [],
                            csgp: [],
                        },
                    },
                    unknown_boxes: [],
                },
                unknown_boxes: [],
                udta: None,
            },
        ],
        mvex: None,
        unknown_boxes: [],
        udta: Some(
            UserDataBox {
                cprt: [],
                tsel: None,
                kind: [],
                strk: [],
                ludt: [],
                unknown_boxes: [
                    UnknownBox {
                        header: BoxHeader {
                            size: Short(
                                385,
                            ),
                            box_type: FourCc(
                                "meta",
                            ),
                        },
                        data.len: 377,
                    },
                ],
            },
        ),
    },
    moof: [],
    mdat: [
        MediaDataBox {
            data.len: 249997,
        },
    ],
    imda: [],
    meta: None,
    styp: [],
    sidx: [],
    ssix: [],
    prft: [],
    unknown_boxes: [
        UnknownBox {
            header: BoxHeader {
                size: Short(
                    8,
                ),
                box_type: FourCc(
                    "free",
                ),
            },
            data.len: 0,
        },
    ],
    mfra: None,
}
